### PR TITLE
Add inference utilities to transform between unconstrained and constrained space

### LINKED ISSUE #1554: Inverse bjiector transformation (from constrained to unconstrained space)
Hi,
Is it possible, with the current interface, to design a function that takes as input the pytree of _constrained_ parameter values and return _unconstrained_ values, given a model with Pyro primitives? For instance, I am looking for a function that does the inverse transform of `constrain_fn()` from `numpyro.infer.util`.

Sorry to ask here but I could not find the answer in the code. So maybe there is no easy implementation of it yet.

### PR DESCRIPTION
**Pull request by aymgal**
This is a proposal to fix #1554

### CODE REVIEW

File: numpyro/infer/util.py [hunk 1]
```diff
@@ -193,6 +193,70 @@ def substitute_fn(site):
...
+    from unconstrained space to constrained space.
+
+    :param model: a callable containing NumPyro primitives.
+    :param tuple model_args: args provided to the model.
+    :param dict model_kwargs: kwargs provided to the model.
+    :param dict params: dictionary of unconstrained values keyed by site
+        names.
+    :return: `dict` of transformation keyed by site names.
+    """
+    transforms = get_transforms(model, model_args, model_kwargs, params)
```
[the remark this answers is not in the document]
Comment (aymgal) [reply numpyro/infer/util.py]: Indeed you're right. Let me know if your points have been addressed by the latest commit :)

File: numpyro/infer/util.py [hunk 2]
```diff
@@ -157,7 +157,8 @@ def transform_fn(transforms, params, invert=False):
     return {k: transforms[k](v) if k in transforms else v for k, v in params.items()}
 
 
-def constrain_fn(model, model_args, model_kwargs, params, return_deterministic=False):
+def constrain_fn(model, model_args, model_kwargs, params, 
+                 include_param_sites=False, return_deterministic=False):
```
Comment (fehiepsi) [inline numpyro/infer/util.py:161]: I think you can remove those `include_param_sites` flags and keep the True behavior by default.

Comment (aymgal) [reply numpyro/infer/util.py]: This was for ensuring backward compatibility. I removed these in the last comit

File: numpyro/infer/util.py [hunk 3]
```diff
@@ -193,6 +197,39 @@ def substitute_fn(site):
...
+    :param model: a callable containing NumPyro primitives.
+    :param tuple model_args: args provided to the model.
+    :param dict model_kwargs: kwargs provided to the model.
+    :param dict params: dictionary of values keyed by site names.
+    :return: `dict` of transformation keyed by site names.
+    """
+    substituted_model = substitute(model, data=params)
+    transforms, _, _, _ = _get_model_transforms(substituted_model, model_args, model_kwargs)
+    return transforms
+
```
Comment (fehiepsi) [inline numpyro/infer/util.py:215]: Could you run `make format` to fix lint issue? I guess you need to a new line here

File: numpyro/infer/util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/infer/util.py]: I see. I guess it won't affect the old behavior. If users provide param sites, your solution will return expected results, while the current master branch will raise an error - so this is an improvement.

**Comment by aymgal** [comment]
@fehiepsi fyi I rebased the commits of this PR as I had to do lots of syntax fixing recently... should be ok now

**Comment by fehiepsi** [comment]
Thanks @aymgal! It's great to have this utility available.

**Review by fehiepsi — APPROVED** [review]

**Pull request merged**